lib/Sema/TypeCheckStmt.cpp

```diff
@@ -2121,15 +2121,11 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
                                                             builderType);
       if (!body)
         return true;
-    } else if (func->hasSingleExpressionBody()) {
-      auto resultTypeLoc = func->getBodyResultTypeLoc();
-      auto expr = func->getSingleExpressionBody();
-
-      if (resultTypeLoc.isNull() || resultTypeLoc.getType()->isVoid()) {
-        // The function returns void.  We don't need an explicit return, no matter
-        // what the type of the expression is.  Take the inserted return back out.
-        body->setFirstElement(expr);
-      }
+    } else if (func->hasSingleExpressionBody() &&
+               func->getResultInterfaceType()->isVoid()) {
+      // The function returns void.  We don't need an explicit return, no matter
+      // what the type of the expression is.  Take the inserted return back out.
+      body->setFirstElement(func->getSingleExpressionBody());
```

Comment:
@DougGregor Do you have any concern about this change?

Reply:
No, this seems like a clear improvement.